### PR TITLE
Target package names

### PR DESCRIPTION
**Pull request by dcoutts**
Much improved support for named packages in the target selector syntax and resolution. Instead of just supporting syntax of a single unadorned package name, all the existing syntax that allow package names are extended to support named packages.

This includes packages from the `extra-packages` but also indirect deps and packages from hackage. The resolution for `extra-packages` and indirect deps should just work, while currently packages outside the project are reported as an error (but we're in a good position now to resolve to packages from hackage if we want).

So this should be a better basis for the target handling for the new install code, since any command should now be able to refer to packages outside of the project (or at least we're a short step from tha now).

This passes the existing target selector integration tests. New tests need to be added to cover the `extra-packages` and indirect deps within the project, and packages outside the project.

This does extend what the existing syntax can refer to so, docs & changelog will need to be updated.

BTW, this PR is designed to be read patch-by-patch, not all in one go. Each change is hopefully comprehensible on its own.

Please include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/#conventions).
* [ ] Any changes that could be relevant to users have been recorded in the changelog.
* [ ] The documentation has been updated, if necessary.
* [x] If the change is docs-only, `[ci skip]` is used to avoid triggering the build bots.

Please also shortly describe how you tested your change. Bonus points for added tests!

### CODE REVIEW

**Comment by 23Skidoo** [comment]
Travis still red... looks like a failure in the CmdRun test.

**Comment by fgaz** [comment]
[The failing one](https://travis-ci.org/haskell-pushbot/cabal-binaries/builds/305551054) is #4644 again

**Comment by dcoutts** [comment]
Is this a prior failing test? Is this one I need to fix?

**Comment by 23Skidoo** [comment]
CI seems to be green now.

File: cabal-install/Distribution/Client/TargetSelector.hs [hunk 1]
```diff
@@ -1104,10 +1116,14 @@ syntaxForm2PackageComponent ps =
           return (TargetComponent pinfoId (cinfoName c) WholeComponent)
         --TODO: the error here ought to say there's no component by that name in
```
Comment (23Skidoo) [inline cabal-install/Distribution/Client/TargetSelector.hs]: Is this TODO still valid?

**Comment by 23Skidoo** [comment]
@dcoutts Do you want to merge this as-is or do you want to write some additional tests before merging? The latter probably makes more sense once the feature is fully implemented.

File: cabal-install/Distribution/Client/TargetSelector.hs [hunk 2]
```diff
@@ -2131,15 +2133,15 @@ tryEach = exactMatches
...
-  InexactMatch _ [x] -> Unambiguous x
-  ExactMatch   _  [] -> error "findMatch: impossible: ExactMatch []"
-  InexactMatch _  [] -> error "findMatch: impossible: InexactMatch []"
-  ExactMatch   _  xs -> Ambiguous True  xs
-  InexactMatch _  xs -> Ambiguous False xs
-
-data MaybeAmbiguous a = None [MatchError] | Unambiguous a | Ambiguous Bool [a]
+  NoMatch _ msgs -> None msgs
+  Match _ _  [x] -> Unambiguous x
+  Match m d   [] -> error $ "findMatch: impossible: " ++ show match'
```
[the remark this answers is not in the document]
Comment (hvr) [reply cabal-install/Distribution/Client/TargetSelector.hs]: I'd vote for a TODO; this won't be the only place benefiting from NonEmpty

**Comment by 23Skidoo** [comment]
I'm merging this now, because the PR looks quite good on the whole and there's a chance it will start bitrotting.

**Comment by sboosali** [comment]
@dcoutts I got here via @hvr 's comment on #5444 about an "enhancement to its [extra-package's] syntax". could you clarify if you have a moment? (I want to port over a few Nix-based and Stack-based projects, which a few different ways of defining additional vendored packages.)

**Pull request merged**